scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/SparkeyUri.scala

```diff
@@ -49,55 +44,43 @@ trait SparkeyUri extends Serializable {
 
 private[sparkey] object SparkeyUri {
   def apply(basePath: String, opts: PipelineOptions): SparkeyUri =
-    if (ScioUtil.isGcsUri(new URI(basePath))) {
-      new GcsSparkeyUri(basePath, opts.as(classOf[GcsOptions]))
-    } else {
+    if (ScioUtil.isLocalUri(new URI(basePath))) {
       new LocalSparkeyUri(basePath)
+    } else {
+      new RemoteSparkeyUri(basePath, opts.as(classOf[GcsOptions]))
     }
+  def extensions: Seq[String] = Seq(".spi", ".spl")
 }
 
 private class LocalSparkeyUri(val basePath: String) extends SparkeyUri {
   override def getReader: SparkeyReader = Sparkey.open(new File(basePath))
-  override private[sparkey] def exists: Boolean =
-    new File(basePath + ".spi").exists || new File(basePath + ".spl").exists
+  override private[sparkey] def exists: Boolean = {
+    SparkeyUri.extensions.map(e => new File(basePath + e)).exists(_.exists)
+  }
 }
 
-private class GcsSparkeyUri(val basePath: String, options: GcsOptions) extends SparkeyUri {
+private class RemoteSparkeyUri(val basePath: String, options: PipelineOptions) extends SparkeyUri {
 
-  def localBasePath: String = sys.props("java.io.tmpdir") + hashPrefix(basePath)
-  private val json: String = new ObjectMapper().writeValueAsString(options)
+  val rfu: RemoteFileUtil = RemoteFileUtil.create(options)
 
   override def getReader: SparkeyReader = {
-    for (ext <- Seq("spi", "spl")) {
-      ScioUtil.fetchFromGCS(gcs, new URI(s"$basePath.$ext"), s"$localBasePath.$ext")
-    }
-    Sparkey.open(new File(s"$localBasePath.spi"))
-  }
-
-  override def exists: Boolean = {
-    val index = GcsPath.fromUri(basePath + ".spi")
-    val log = GcsPath.fromUri(basePath + ".spl")
-    Try(gcs.fileSize(index)).isSuccess || Try(gcs.fileSize(log)).isSuccess
+    val uris = SparkeyUri.extensions.map(e => new URI(basePath + e))
+    val paths = rfu.download(uris.asJava).asScala
+    Sparkey.open(paths.head.toFile)
   }
-
-  @transient private[sparkey] lazy val gcs: GcsUtil = new ObjectMapper()
-    .readValue(json, classOf[PipelineOptions])
-    .as(classOf[GcsOptions])
-    .getGcsUtil
-
-  private def hashPrefix(path: String): String =
-    "sparkey-" + Hashing.sha1().hashString(path, Charsets.UTF_8).toString.substring(0, 8)
-
+  override private[sparkey] def exists: Boolean =
+    SparkeyUri.extensions
+      .exists(e => rfu.remoteExists(new URI(basePath + e)))
 }
 
 private[sparkey] class SparkeyWriter(val uri: SparkeyUri) {
 
   private val localFile = uri match {
-    case u: LocalSparkeyUri => u.toString
-    case u: GcsSparkeyUri => u.localBasePath
+    case u: LocalSparkeyUri => u.basePath
+    case _: RemoteSparkeyUri => Files.createTempDirectory("sparkey").resolve("data").toString
```

Comment:
So the hash prefix is no longer needed right? This should be unique every time it's called / internally uses some kind of hash in the name?